### ISSUE TITLE
Incorporate "predictor.py" into "smite --trace".

### ISSUE THREAD

**Issue by apt1002** [rrthomas/smite#42]
Building a predictor (using the "intrumentented_apt" branch) is currently a two-step process:

1. Record a trace file using "smite --trace".
2. Replay it and summarise it using the "predictor.py" script.

The trace file that is passed from step 1 to step 2 is often huge, and reading and writing it is a significant expense. It would be more efficient to rewrite "predictor.py" in C, and treat it as a lossy compression algorithm that it unconditionally applied to the output of "smite --trace". In other words, deprecate trace files in favour of predictor files.

**Comment by rrthomas** [comment]
I am leery of adding more complexity to SMite's front-end in this regard. I just tried setting the `--trace` parameter to a named pipe and having `gen-predictor` read from that. It still seems to be pretty slow (definitely slower than writing the file). However, presumably that's because `gen-predictor` is slow? If I simply pipe the output into zstd, it takes ~100s to build pforth (zstd can't read from a fifo, so I have to `cat fifo | zstd - -o trace.zstd`.

I could certainly have smite respect the convention of accepting `-` to mean "standard output" for the filename, so that normal (anonymous) pipes could be used.

**Comment by apt1002** [comment]
Yes, "gen-predictor" is much slower than "smite --trace", because Python.

It is extra complexity, and I cannot deny it, however well I can compartmentalize it. I'm certainly still at a point where "do nothing" is feasible.

I have not profiled "smite --trace", but there is only one line that could explain the slowdown, and it's a call to `fprintf()`. That serialization code, and the matching deserialization code in "gen-predictor", is certainly going to be slower than the main job that "gen-predictor" does, which is very simple.

The complexity
---

Let me enumerate the new bits, and the bits we could delete, so we know what we're talking about.

 - In `trace()`:
```
predictor.record_event(history, event)
history = step(history, event)
```
 - `record_event()` does the following, which could be simplified if we used an array instead of a hashtable, because 70% of it is concerned with things happening for the first time:
```
if history not in self.predictions:
    self.predictions[history] = {}
counts = self.predictions[history]
if event.name not in counts:
    counts[event.name] = 0
counts[event.name] += 1
```
 - `step()` does this magically simple calculation:
```
or_mask = event.hash0 & event.hash1
and_mask = event.hash0 | event.hash1
return (history | or_mask) & and_mask
```
 - Some calls in main to initialise the predictor data structure and to write it out at the end.
 - A predictor file format, because pickle would not be available.
 - Code in "gen-labels" to read the predictor file format.

I think that's the complete list of additions. On the other side of the scale:

 - The algorithm is already in "predictor.py" (albeit in a much more concise language), and could be removed from there.

 - We can forget the concept of a trace file completely.

**Comment by rrthomas** [comment]
Now that the trace function is entirely in main.c, changing it would be less of a burden. But I would still favour a separate C program into which the trace is piped; does that sound plausible?

**Comment by apt1002** [comment]
Yes, it sounds plausible.

**Comment by rrthomas** [comment]
Update: with the new features system, rewriting `trace_run` would be fine.

**Comment by rrthomas** [comment]
@apt1002 writes:

Progress: I changed StepFunction. Old one was based on OR and AND. New one is based on OR and XOR. This means that if you repeat the same steps, each iteration toggles the history between two values, thus remembering whether the number of iterations is even or odd. It makes a lot more states: 6x. However, trace is 48KB.

I've now achieved 10KB with a 20-bit history (down from 63-bit) and even more states. zstd achieves 3202 bytes, so I'm not even a big boy yet, but it's an improvement. The significance of using a 20-bit history is that I no longer need hashtables, so I can rewrite it in C.

**Comment by apt1002** [comment]
gen-predictor is now written in C (on branch "specializer") which is a step towards closing this issue 

A digression about the prediction algorithm follows.

The OR/XOR prediction algorithm change did make `mit -O` slightly faster.

The faster gen-predictor allows longer instruction traces to he used, which requires retuning. I reduced the number of predictor states, by making the history more forgetful. This reduced prediction accuracy on the shorter trace a bit, but did not slow down `mit -O` measurably.

I did experiments with a longer trace (`pForth make-base.fs`). The longer trace did not result in a faster `mit -O`.

I conclude that the prediction accuracy is not particularly limiting at the moment. The current prediction algorithm will do for now.
